pkg/rtc/participant_internal_test.go

```diff
@@ -176,6 +176,32 @@ func TestTrackPublishing(t *testing.T) {
 		// check SID is the same
 		require.Equal(t, p.pendingTracks["cid"].trackInfos[0].Sid, p.pendingTracks["cid"].trackInfos[1].Sid)
 	})
+
+	t.Run("should not allow adding disallowed sources", func(t *testing.T) {
+		p := newParticipantForTest("test")
+		p.SetPermission(&livekit.ParticipantPermission{
+			CanPublish: true,
+			CanPublishSources: []livekit.TrackSource{
+				livekit.TrackSource_CAMERA,
+			},
```

Comment:
I forget now, but is it required to set `CanPublish` always and `CanPublishSources` is an additional layer or setting some source means `CanPublish` is not required? From code changes in `participant.go` ☝️ , look like `CanPublish` is not checked, but don't remember if that is abstracted in grants check routines. If there are some combinations here and if only some are allowed, it would be good to add a test for those combinations. For example, if `CanPublish` is required, then would be good to set that to `false`, but set some can publish sources and ensure that they still do not go through.

Reply:
it is not required to set CanPublish in the grant itself. When it's nil, it'll be allowed.

I'm setting it explicitly here because when you update permissions, none of the fields are nil-able. So you'd have to explicitly set it, otherwise CanPublish would be set to `false`

Reply:
the logic that enforces these rules are [here](https://github.com/livekit/protocol/blob/main/auth/grants.go#L92).